### DISCUSSION TITLE
issues in executing code for Dynamic Phase Field Fracture using Explicit Time Integration Scheme.

### DISCUSSION THREAD

**Discussion by zuckarm** [idaholab/moose#28623] [Q&A Modules: Phase field]
hi all,

i am working on a dynamic fracture problem: a 2d notched plate is subjected to traction on its top and bottom edges. i am using the MultiApps system to separate the solid mechanics problem in the main app, and phase field problem in the sub app. i want to solve the problem using explicit time integration scheme.

i am facing the following errors when i run the code:
```
Time Step 0, time = 0
Currently Executing
  Finished Performing Initial Setup                                                      [  7.32 s] [  460 MB]

Time Step 0, time = 0

Time Step 1, time = 5e-09, dt = 5e-09
Initial fixed point residual norm: 0.000000e+00

Beginning fixed point iteration 0


Main app solve:

Performing automatic scaling calculation

The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
 Solve Did NOT Converge!
  Finished Solving                                                                       [  9.26 s] [  813 MB]
Fixed point convergence reason: DIVERGED_NONLINEAR
Aborting as solve did not converge

```

i have tried a number of edits to the code, but the initial fixed point residual norm is always reported as zero, whether i use an explicit or implicit scheme. 
while using an implicit scheme, the nonlinear residuals are zero at all timesteps, due to which the code progresses but it reports values of all variables: displacement, phase field as zero for all timesteps.

can anyone help me in resolving this issue?

problem data:
[upload.zip](https://github.com/user-attachments/files/17006183/upload.zip)

**Comment by GiudGiud** [comment]
Hello

is this transfer in the right direction?
```
	
	[./to_psie_active]
		type			= MultiAppCopyTransfer
		from_multi_app 	= 'fracture'
		variable		= psie_active
		source_variable = psie_active
	[../]
```
shouldnt it be to_multi_app?

You should make sure both inputs are valid with aritifical data from the other input (setting the transferred auxiliary variables using initial conditions for example)

**Reply by zuckarm** [reply]
yes, it must be "to_multi_app". i made the changes, but i am still facing same error as before.
i have changed preconditioning too, but i am still getting initial residual as zero. 
what could be the reason for initial residual to be zero?

**Reply by zuckarm** [reply]
i have made a few more changes, and now the code seems to work fine for implicit scheme. however, when i go for explicit scheme, i get the following error:
`Initial fixed point residual norm: inf`
i have tried reducing the timestep, but to no avail. can anyone please suggest any alternatives?

**Reply by GiudGiud** [reply]
you have to go kernel by kernel (and BCs too) and check which one is returning the inf number
This interface, which kernels use, will let you save their contribution to the residual in an auxiliary variable
https://mooseframework.inl.gov/source/interfaces/TaggingInterface.html

**Reply by zuckarm** [reply]
hi, is there any example input file which shows how to use the TaggingInterface?

**Reply by GiudGiud** [reply]
see this folder
http://github.com/idaholab/moose/tree/next/test/tests/tag

**Reply by zuckarm** [reply]
hi, just wanted to make sure that i understand correctly-
1. TaggingMatrixAux takes in an AuxVariable, and appends the AuxVariable with the sum of diagonal values of the tagging matrix at each time step, right?
2. i wanted to do something like create a matrix at each time step, and extract individual components of the matrix by using a rank two tensor. i wanted to do the matrices for the kernels and materials. and maybe vectors for the boundary conditions, etc. how can i do that using TaggingInterface system?
3. the point of doing this is because i am formulating a dynamic phase field fracture problem using two-way coupling, and i am seeing that the solver does not progress beyond the first time step and throws an error like diverged nonlinear iterations. as per your earlier suggestions, i want to understand what is happening in the individual matrices and vectors, and whether the final system that we solve- A*x=b is being implemented correctly.
4. as far as i understand, this A matrix depends on the problem formulation, but majorly it depends on the Executioner system parameters. by extracting matrix-and vector-forms of the contributing systems, i want to just keep changing the Executioner parameters and see what happens to the final A matrix. is it possible to do so by using the TaggingInterface?

**Reply by GiudGiud** [reply]
> the point of doing this is because i am formulating a dynamic phase field fracture problem using two-way coupling, and i am seeing that the solver does not progress beyond the first time step and throws an error like diverged nonlinear iterations. as per your earlier suggestions, i want to understand what is happening in the individual matrices and vectors, and whether the final system that we solve- A*x=b is being implemented correctly.

for that, just use `-ksp_view_mat` / `-ksp_view_pmat` on the command line. This will output the system matrix so you can check it


> as far as i understand, this A matrix depends on the problem formulation, but majorly it depends on the Executioner system parameters. by extracting matrix-and vector-forms of the contributing systems, i want to just keep changing the Executioner parameters and see what happens to the final A matrix. is it possible to do so by using the TaggingInterface?

It is possible to do so. But see the petsc options, it's easier that way


> TaggingMatrixAux takes in an AuxVariable, and appends the AuxVariable with the sum of diagonal values of the tagging matrix at each time step, right?

No the local value of the diagonal of the tagging matrix. There is only tagged matrix per TaggingMatrixAux, which defaults to the Jacobian / its diagnoal approximation

> i wanted to do something like create a matrix at each time step, and extract individual components of the matrix by using a rank two tensor. i wanted to do the matrices for the kernels and materials. and maybe vectors for the boundary conditions, etc. how can i do that using TaggingInterface system?

You can add new tagged matrices. In kernels and BCs you can set the matrix tags parameter to do that. in a material you would need to do it programmatically.

The way kernels and BCs contribute to tagged matrices is fixed, it's the contribution to the Jacobian. If you want to change that, and contribute something else, you ll need to do it programatically  by overriding computeJacobian() in the kernel for example.

>  and extract individual components of the matrix by using a rank two tensor

where is this rank two tensor going to live? As a material property? As 4 independent auxiliary variables?

**Reply by zuckarm** [reply]
what do you mean by local value of diagonal of tagging matrix? 

about rank two tensor- i wanted to extract certain components of the matrices formed, what is the way to do so?

**Reply by GiudGiud** [reply]
The contribution to the Jacobian is performed on every element. That's a matrix that is used by default in the tagging system for kernels. You can access these values on every element, that's what I mean by local.

**Reply by zuckarm** [reply]
okay, thank you.
is there any way to store -ksp_view_mat output to a csv file? i can see the matrices on the console.

**Reply by GiudGiud** [reply]
yes there is an option to dump the output to a matlab file.
see petsc manual for it

**Reply by zuckarm** [reply]
thank you, i am able to extract the data to a MATLAB file, however, when i do so for a problem involving explicit time integration scheme like central difference, i am seeing the following error:

`The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.
A libMesh::PetscSolverException was raised during FEProblemBase::computeJacobianTags
Not for unassembled matrix`

to use explicit time integators, i have made `solve_type = LINEAR`

**Reply by GiudGiud** [reply]
for an explicit solve, we dont form the Jacobian.

we do form a mass matrix I think ? I am not sure how to access that.
@lindsayad you would know from your recent PR :)

**Reply by lindsayad** [reply]
What option are you passing to extract the data to Matlab?

**Reply by zuckarm** [reply]
i used `petsc_options = '-ksp_view_mat'`

i had another question --
i am working on phase field fracture which uses the MultiApp system. is there any way to get a coupled matrix for the two problems?

**Reply by zuckarm** [reply]
i am solving a dynamic phase field fracture problem using explicit time integration scheme, and i am facing the following error:

```
Main app solve:

Performing automatic scaling calculation

The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
The 1-th row of A is exactly zero
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
 Solve Did NOT Converge!
  Finished Solving                                                                       [ 13.90 s] [  617 MB]
Fixed point convergence reason: DIVERGED_NONLINEAR
Aborting as solve did not converge
```

this is the executioner block in the input file:
```
[Executioner]
  type        = Transient
  solve_type  = LINEAR

  automatic_scaling = true

  petsc_options       = '-snes_converged_reason -ksp_converged_reason -pc_svd_monitor'
  petsc_options_iname = '-pc_type   -pc_factor_mat_solver_package   -pc_factor_shift_type'
  petsc_options_value = 'lu         superlu_dist                    NONZERO'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  nl_max_its = 50

  l_tol       = 1e-6
  l_max_its   = 10
  line_search = 'none'

  dt        = 6e-8
  end_time  = 80e-6

  fixed_point_max_its = 20

  fixed_point_rel_tol = 1e-8
  fixed_point_abs_tol = 1e-10

  [TimeIntegrator]
    type = CentralDifference
  []
[]
```

can anyone please help me resolve this error? thanks.

**Reply by GiudGiud** [reply]
> i am working on phase field fracture which uses the MultiApp system. is there any way to get a coupled matrix for the two problems?

no you cannot get the coupling matrix. The MultiApp sets up a fixed point iteration algorithm, not a Newton-like method with a Jacobian where you could see the coupling terms

> i am solving a dynamic phase field fracture problem using explicit time integration scheme, and i am facing the following error:

can you attach the whole input?
I think you would get this from missing the time derivative term

**Reply by zuckarm** [reply]
main file:

```
# 'dynamic phase field fracture'

E   = 32e3
nu  = 0.2
K   = '${fparse E/3/(1-2*nu)}'
G   = '${fparse E/2/(1+nu)}'
rho = 2.45e-6

Gc  = 3e-6
l   = 0.25

[MultiApps]
  [fracture]
    type          = TransientMultiApp
    input_files   = 'crack_branch_sub.i'
    cli_args      = 'Gc=${Gc};l=${l}'
    execute_on    = 'TIMESTEP_END'
  []
[]

[Transfers]
  [from_d]
    type              = MultiAppCopyTransfer
    from_multi_app    = 'fracture'
    variable          = d
    source_variable   = d
  []

  [to_psie_active]
    type              = MultiAppCopyTransfer
    to_multi_app      = 'fracture'
    variable          = psie_active
    source_variable   = psie_active
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Mesh]
  type  = FileMesh
  file  = 'crack_branch.msh'

[Variables]
  [disp_x]
    order   = FIRST
    family  = LAGRANGE
  []

  [disp_y]
    order   = FIRST
    family  = LAGRANGE
  []
[]

[AuxVariables]
  [react_y]
    order   = FIRST
    family  = LAGRANGE
  []

  [d]
  []

  [accel_x]
  []

  [accel_y]
  []

  [vel_x]
  []

  [vel_y]
  []
[]

[Kernels]
  [dynamic_x]
    type      = ADDynamicStressDivergenceTensors
    component = 0
    variable  = disp_x
  []

  [dynamic_y]
    type      = ADDynamicStressDivergenceTensors
    component = 1
    variable  = disp_y
    save_in   = react_y
  []

  [inertia_x]
    type          = ADInertialForce
    variable      = disp_x
    density       = rho
  []

  [inertia_y]
    type          = ADInertialForce
    variable      = disp_y
    density       = rho
  []
[]

[AuxKernels]
  [acceleration_x]
    type          = TestNewmarkTI
    variable      = accel_x
    displacement  = disp_x
    first         = false
  []

  [acceleration_y]
    type          = TestNewmarkTI
    variable      = accel_y
    displacement  = disp_y
    first         = false
  []

  [velocity_x]
    type          = TestNewmarkTI
    variable      = vel_x
    displacement  = disp_x
    first         = true
  []

  [velocity_y]
    type          = TestNewmarkTI
    variable      = vel_y
    displacement  = disp_y
    first         = true
  []
[]

[BCs]
  [load_top]
    type      = Pressure
    boundary  = top
    factor    = -1
    variable  = disp_y
  []

  [load_bottom]
    type      = Pressure
    boundary  = bottom
    factor    = -1
    variable  = disp_y
  []
[]

# [ICs]
#   [init_phase_field]
#     type      = BrittleDamageIC
#     d0        = 1
#     l         = 0.25
#     variable  = d
#     x1        = 0
#     x2        = 50
#     y1        = 20
#     y2        = 20
#     z1        = 0
#     z2        = 0
#   []
# []

[Materials]
  [bulk_props]
    type        = ADGenericConstantMaterial
    prop_names  = 'K    G     rho'
    prop_values = '${K} ${G}  ${rho}'
  []

  [degradation_func]
    type              = PowerDegradationFunction
    property_name     = g
    expression        = '(1-d)^p*(1-eta)+eta'
    phase_field       = d
    parameter_names   = 'p  eta'
    parameter_values  = '2  1e-6'
  []

  [elastic_props]
    type                  = SmallDeformationIsotropicElasticity
    bulk_modulus          = K
    shear_modulus         = G
    phase_field           = d
    degradation_function  = g
    decomposition         = SPECTRAL
    output_properties     = 'elastic_strain psie_active'
    outputs               = exodus
  []

  [strain_calc]
    type  = ADComputeSmallStrain
  []

  [stress_calc]
    type              = ComputeSmallDeformationStress
    elasticity_model  = elastic_props
    output_properties = 'stress'
    outputs           = exodus
  []
[]

[Postprocessors]
  [reaction]
    type      = NodalSum
    variable  = react_y
    boundary  = top
  []

  [phase_field_var]
    type      = AverageNodalVariableValue
    variable  = d
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type        = Transient
  solve_type  = LINEAR

  automatic_scaling = true

  petsc_options       = '-snes_converged_reason -ksp_converged_reason -pc_svd_monitor'
  petsc_options_iname = '-pc_type   -pc_factor_mat_solver_package   -pc_factor_shift_type'
  petsc_options_value = 'lu         superlu_dist                    NONZERO'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  nl_max_its = 50

  l_tol       = 1e-6
  l_max_its   = 10
  line_search = 'none'

  dt        = 6e-8
  end_time  = 80e-6

  fixed_point_max_its = 20

  fixed_point_rel_tol = 1e-8
  fixed_point_abs_tol = 1e-10

  [TimeIntegrator]
    type = CentralDifference
  []
[]

[Outputs]
  exodus    = true
  csv       = true
  file_base = mar08/crack_branch
[]
```
sub file:

```
[Mesh]
  type = FileMesh
  file = 'crack_branch.msh'
[]

[Variables]
  [d]
  []
[]

[AuxVariables]
  [bounds_dummy]
  []

  [psie_active]
    order   = CONSTANT
    family  = MONOMIAL
  []
[]

[Bounds]
  [irreversibility]
    type              = VariableOldValueBounds
    variable          = bounds_dummy
    bounded_variable  = d
    bound_type        = lower
  []

  [upper]
    type              = ConstantBounds
    variable          = bounds_dummy
    bounded_variable  = d
    bound_type        = upper
    bound_value       = 1
  []
[]

[Kernels]
  [diffusion]
    type                    = ADPFFDiffusion
    variable                = d
    fracture_toughness      = Gc
    regularization_length   = l
    normalization_constant  = c0
  []

  [source]
    type        = ADPFFSource
    variable    = d
    free_energy = psi
  []
[]

[Materials]
  [fracture_props]
    type        = ADGenericConstantMaterial
    prop_names  = 'Gc     l'
    prop_values = '${Gc}  ${l}'
  []

  [degradation_func]
    type              = PowerDegradationFunction
    property_name     = g
    expression        = '(1-d)^p*(1-eta)+eta'
    phase_field       = d
    parameter_names   = 'p eta'
    parameter_values  = '2 1e-6'
  []

  [crack_geom_func]
    type          = CrackGeometricFunction
    property_name = alpha
    expression    = 'd^2'
    phase_field   = d
  []

  [psi]
    type                    = ADDerivativeParsedMaterial
    property_name           = psi
    expression              = 'alpha*Gc/c0/l+g*psie_active'
    coupled_variables       = 'd  psie_active'
    material_property_names = 'alpha(d) g(d) Gc c0 l'
    derivative_order        = 1
  []
[]

[Executioner]
  type        = Transient
  solve_type  = LINEAR

  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
[]
```

**Reply by GiudGiud** [reply]
ah I need crack_branch.msh to run this. But still, I looked at your input and
you are missing a time derivative kernel. Since you dont have anything contributing to the mass matrix, you may not use the CentralDifference time integrator

**Reply by zuckarm** [reply]
i am adding DynamicStressDivergenceTensors and InertialForce kernels in the main file. there is no time dependence in the sub file. where should i add a time derivative kernel? also, do i need to add a mass matrix?

**Reply by GiudGiud** [reply]
In the kernels block you can add a time derivative. 

I'm not sure you should be using an explicit time integrator for a pseudo transient with load steps rather than time steps. So is your simulation a real transient where you want to be looking at the evolution of the variables through time?

**Reply by lindsayad** [reply]
If you update your MOOSE, you would get an error for that input

**Reply by zuckarm** [reply]
my simulation is a real transient, i want to see the evolution of variables in both main and sub file through time.
is a separate time derivative kernel necessary when DynamicStressDivergenceTensors and InterialForce kernels are already added?

**Reply by GiudGiud** [reply]
I'm fairly sure you need time derivatives if not using the DynamicSolidMechanicsPhysics.
@dschwen pls correct me if I'm wrong though
